File: users/chenjiasheng/547. 省份数量.py
from typing import List


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        parents = list(range(n))
        sizes = [1] * n

        def find(x):
            if parents[x] != x:
                parents[x] = find(parents[x])
            return parents[x]

        def union(x, y):
            x, y = find(x), find(y)
            if x == y:
                return
            if sizes[x] < sizes[y]:
                x, y = y, x
            parents[y] = x
            sizes[x] += sizes[y]
            sizes[y] = 0

        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j]:
                    union(i, j)

        result = 0
        for i in range(n):
            if parents[i] == i:
                result += 1

        return result


isConnected = [[1,0,0],[0,1,0],[0,0,1]]
solution = Solution()
result = solution.findCircleNum(isConnected)
print(result)

File: users/xiewufeng/547. 省份数量.cpp
class UnionFind {
public:
    vector<int> parents;
    UnionFind(int size)
    {
        parents.resize(size);
        for (int i = 0; i < size; i++)
            parents[i] = i;
    }

    int find (int idx)
    {
        if (parents[idx] != idx)
            parents[idx] = find(parents[idx]);
        return parents[idx];
    }

    void unite(int idx1, int idx2)
    {
        int rootX = find(idx1);
        int rootY = find(idx2);
        parents[rootX] = rootY;
    }
};

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        UnionFind uf(n);
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                if ( isConnected[i][j] == 1 )
                    uf.unite(i, j);
            }
        }

        int count = 0;
        for (int k = 0; k < n; k++) {
            if (uf.find(k) == k)
                count++;
        }

        return count;

    }
};